New PSR for ClockInterface (#1224)
php-fig/fig-standards: proposed/clock.md

Common Interface for Accessing the Clock
========================================

This document describes a simple interface for reading the system clock.

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
"SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be
interpreted as described in [RFC 2119][].

The final implementations MAY decorate the objects with more
functionality than the one proposed but they MUST implement the indicated
interfaces/functionality first.

[RFC 2119]: http://tools.ietf.org/html/rfc2119

# 1. Specification

## 1.1 Introduction

Creating a standard way of accessing the clock, would allow interopability
during testing, when testing behavior that has timing based side affects.
Common ways to get the current time include calling `\time()` or 

> `new DateTimeImmutable('now')` however this makes mocking the current time

Reviewer: ```suggestion
`new DateTimeImmutable('now')`. However, this makes mocking the current time
```

impossible in some situations.

## 1.2 Definitions

* **Clock** - The clock is able to read the current time and date.

* **Timestamp** - The current time as an integer number of seconds since
Jan 1, 1970 00:00:00 UTC.

### 1.3 Usage

There are some common usage patterns, which are outlined below:

**Get the current timestamp**

This should be done by using the `getTimestamp()` method on the returned `\DateTimeImmutable` like so:
```php
$timestamp = $clock->now()->getTimestamp();
```

> **Timezone**

Reviewer 2: Why are timezones a usage pattern?

A usage pattern for a ClockInterface should only be to get the current date and time via `now()`. Everything else (getting the timesamp, getting a specially formatted DateTime-string etc) is then calling appropriate functionality on the returned Object.

> Each implmenetation of the `ClockInterface` is free to return the time in the 

Reviewer 3: Just a small spelling mistake.
```suggestion
Each implementation of the `ClockInterface` is free to return the time in the 
```

> timezone of that library authors choice. This could include but not be limited 

Reviewer 4: ```suggestion
timezone of that library author's choice. This could include but not be limited 
```

> to return the current PHP timezone (as DateTimeImmutable consructor currently

Reviewer 5: ```suggestion
to return the current PHP timezone (as the `DateTimeImmutable` constructor currently
```

> to return the current PHP timezone (as the `DateTimeImmutable` constructor currently

Reviewer 4: ```suggestion
to returning the current PHP timezone (as the `DateTimeImmutable` constructor currently
```

> does), return a timezone set a the creation of the ClockInterface implementation

Reviewer 6: ```diff
- set a the creation
+ set at the creation
```

> does), return a timezone set a the creation of the ClockInterface implementation

Reviewer 5: ```suggestion
does), return a timezone set at the creation of the `ClockInterface` implementation
```

> instance, or always returning a fixed timzone (eg UTC).

Reviewer 7: ```suggestion
instance, or always returning a fixed timezone (e.g. UTC).
```

# 2. Interfaces

## 2.1 ClockInterface

The clock interface defines the most basic operations to read the current time and date from the clock. 

> It MUST return the current time as a DateTimeImmutable.

Reviewer 5: ```suggestion
It MUST return the time as a `DateTimeImmutable`.
```

As others have pointed out, it does not need to be the current time, for example, when using a `FrozenClock` for test purposes. It returns a time, that is it.

Author: It still the current time, in the context of the code that is being excercised when mocking for a test.

~~~php
<?php

namespace Psr\Clock;

interface ClockInterface
{
    /**
     * Returns the current time as a DateTimeImmutable Object
     */

>     public function now(): \DateTimeImmutable;

Reviewer 8: Why `now()`?
We also use Clock interfaces in our projects (different versions) and we discussed what's a good method name.
We decided to use `read()`, a _verb_ which properly describes the intent, and it's also easier on the eyes:
`$clock->read();` instead of `$clock->now();`

Also, technically, `now()` can be somewhat ambiguous. Depending on the implementation, it may not return _now_, but some datetime in the past or future.

Reviewer 9: Great observation! In fact you could intentionally have an instance of a clock which does represent something that is not "now", so reading is a great word to use!

Author: Yes, the working group will have to discuss this and the interface name and the rest of the spec. However now() is used by the three top three libs on packaging, so there may be some weight to being drop in compatible there.

Reviewer 10: > However now() is used by the three top three libs on packagist, so there may be some weight to being drop in compatible there.

Based on what's stated on the repos, I'm the one to blame for using `now()` since folks basically forked my lib to expand the range of supported versions 😅

We'll indeed discuss this on the working group but to publicly state my rationale on this: I've based my implementation on ideas I got from the Java API ([Clock](https://docs.oracle.com/javase/8/docs/api/java/time/Clock.html) and [Instant](https://docs.oracle.com/javase/8/docs/api/java/time/Instant.html)), just decided to simplify things.

> We decided to use read(), a verb which properly describes the intent

We'll take your input to the WG. My 2 cents is that `read()` is a too broad verb and it requires more context: `Hey clock, read` what comes to my mind is "okay, read what?". I do believe we can find a better word to clarify the intention here since `now()` is not even a verb and might indeed be vague (`Hey clock, give me now` - "give me whaaat!?").

Reviewer 4: I'd argue for now() simply because it seems to be the most ubiquitous in usage across projects.  `kreait/clock-php`, `ergebnis/clock`, and `lcobucci/clock` use 'now'.  In other languages 'now' is also widely used, eg Go and Rust.

Reviewer 11: Something else to consider is whether the user of the clock should assume it shows the current time or not. i.e. What is the intent of the interface? For example if the main use case is mocking in tests, then yes the mocked clock would be returning a past/future date probably, but to the *user* of the clock, the intent is that it wants the current time, and it should use `now()`'s return value as if it was indeed the current time. 

I can't really see another case than this which would not break interoperability. Like if I say I need a ClockInterface object but please give me one that's 5 days in the future.. That's not clear at all from the interface perspective anymore, so it breaks assumptions/interop.

I hope this makes sense, not sure if I got my point across but if I keep explaining further it'll start sounding like a time travel film so I'll stop here.

Reviewer 12: I guess, this should just be tested using:
```php
$clock instanceof MockClockInterface
```
with `MockClockInterface` provided by PSR to flag any clock that is supposed to falsify the time and so should be implemented by user-land clock used to mock.

Author: @kylekatarnls that sounds like in instant testing mess, howe do you test the branch thats expecting the actual wall clock time, vs the Mock clock time.

@Seldaek to get the future time you could just make a wrapper class like so:
```php
class FutureTime {
  public function __construct(private ClockInterface $clock, private DateInterval $offset){}

   public function getTime():DateTimeImmutable {
    return $this->clock->now()->add($offset);
  }
}
I dont think there should be any expectation of the `ClockInterface` to provide that sort of functionality.

Author: We are sticking to `now()` as it is consistent with other references to now in PHP, and also in a lot of other libraries as mentioned above.

interpreted as described in [RFC 2119][].

The final implementations MAY decorate the objects with more
functionality than the one proposed but they MUST implement the indicated
interfaces/functionality first.

[RFC 2119]: http://tools.ietf.org/html/rfc2119

# 1. Specification

## 1.1 Introduction

Creating a standard way of accessing the clock, would allow interopability
during testing, when testing behavior that has timing based side affects.
Common ways to get the current time include calling `\time()` or 
`new DateTimeImmutable('now')` however this makes mocking the current time
impossible in some situations.

## 1.2 Definitions

* **Clock** - The clock is able to read the current time and date.

* **Timestamp** - The current time as an integer number of seconds since
Jan 1, 1970 00:00:00 UTC.

# 2. Interfaces

## 2.1 ClockInterface

The clock interface defines the most basic operations to read the current time and date from the clock. 
It MUST return the current time as a DateTimeImmutable, timestamp integer and timestamp float, 
and MUST return the current time zone.

~~~php
<?php

namespace Psr\Clock;

> interface ClockInterface

Reviewer 5: Do we need an `Interface` suffix here?

Author: I am following the style of the other PSR's, like PSR-7 where it has MessageInterface, RequestInterface, etc. Actually his covers it: https://www.php-fig.org/bylaws/psr-naming-conventions/

{
    /**

>      * Reads the current time as a unix timestamp

Reviewer 5: ```suggestion
     * Returns the current time as a unix timestamp
```

>      *
>      * @return int The current time as a unix timestamp

Reviewer 5: ```suggestion
```

     */
    public function timestamp(): int;
    
    /**

>      * Reads the current time as a DateTimeImmutable Object

Reviewer 5: ```suggestion
     * Returns the current time as a DateTimeImmutable Object
```

>      *
>      * @return \DateTimeImmutable The current time as a DateTimeImmutable Object

Reviewer 5: ```suggestion
```

     */
    public function now(): \DateTimeImmutable;

    /**
     * Reads the current time as with microsecond resolution
     *
     * @return float The current time as a float with microsecond resolution
     */

>      public function microtime(): float;

Reviewer 5: If we add `microtime()` here, do we also need `hrtime()`?

See https://www.php.net/manual/en/function.hrtime.php.

Personally, I would limit the interface to

```
<?php

interface Clock
{
    public function now(): \DateTimeImmutable
}
```

So far I haven't found any cases where I need more than that.

Reviewer 7: I definitely agree with this.

Author: I have updated the ClockInterface to only include a single method `now()`

     
    /**
     * Retreieves the current time zone
     *
     * @return \DateTimeZone The current timezone as a DateTimeZone object
     */

>      public function timezone(): \DateTimeZone;

Reviewer 5: The `DateTimeZone` is already composed into the `DateTimeImmutable` object returned from `now()`. 

Do we really need an additional accessor here? What problem would that solve?

Common Interface for Accessing the Clock
========================================

This document describes a simple yet extensible interface for a cache item and

> a cache driver.

Reviewer 13: Maybe a typo in the description?

Author: Thanks for proof-reading.
